### ISSUE TITLE
try-rope-embedding

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/cerebros-core-algorithm-alpha#166]

**Comment by david-thrower** [comment]
hold/do-not-merge: Reasoning:

The implementation may be theoretically mathematically correct, but is a remarkably inefficient implementation. Defer to: #167 which is training much faster and at prima fascie appears to be getting slightly higher accuracy.

- Vectorization: Use TensorFlow's optimized functions to reduce Python-level overhead.
- Precompute Sigmoid/Softmax: Precompute necessary operations outside the call method to avoid repetitive computations during training.
- Use einsum: Leverage einsum for batched matrix multiplications to vectorize operations and reduce element-wise loops.

**Comment by david-thrower** [comment]
Included in #172
